### PR TITLE
bin/acpp: Forbid use of -fyscl flags

### PR DESCRIPTION
**Pull request by al42and**

### CODE REVIEW

**Comment by al42and** [comment]
I opted for a warning, but it might also be a fatal error. I checked with hip, cuda, and generic targets, and in all three cases the errors are equally cryptic, like this:

```console
$ acpp -O3 -DU_SIMPLE -fsycl -c main.cpp
In file included from <built-in>:1:
In file included from /cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_runtime_wrapper.h:115:
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:292:29:
 error: C-style cast from 'int *' to '__attribute__((address_space(5)))
int *' converts between mismatching address spaces
      __ocml_frexp_f32(__x, (__attribute__((address_space(5))) int *)&__tmp);
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:383:28:
 error: C-style cast from 'float *' to
'__attribute__((address_space(5))) float *' converts between mismatching
 address spaces
      __ocml_modf_f32(__x, (__attribute__((address_space(5))) float *)&__tmp);
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:485:17:
 error: C-style cast from 'int *' to '__attribute__((address_space(5)))
int *' converts between mismatching address spaces
      __x, __y, (__attribute__((address_space(5))) int *)&__tmp);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:544:30:
 error: C-style cast from 'float *' to
'__attribute__((address_space(5))) float *' converts between mismatching
 address spaces
      __ocml_sincos_f32(__x, (__attribute__((address_space(5))) float *)&__tmp);
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:555:12:
 error: C-style cast from 'float *' to
'__attribute__((address_space(5))) float *' converts between mismatching
 address spaces
      __x, (__attribute__((address_space(5))) float *)&__tmp);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:870:29:
 error: C-style cast from 'int *' to '__attribute__((address_space(5)))
int *' converts between mismatching address spaces
      __ocml_frexp_f64(__x, (__attribute__((address_space(5))) int *)&__tmp);
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:957:28:
 error: C-style cast from 'double *' to
'__attribute__((address_space(5))) double *' converts between
mismatching address spaces
      __ocml_modf_f64(__x, (__attribute__((address_space(5))) double *)&__tmp);
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:1051:17:
 error: C-style cast from 'int *' to '__attribute__((address_space(5)))
int *' converts between mismatching address spaces
      __x, __y, (__attribute__((address_space(5))) int *)&__tmp);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:1112:12:
 error: C-style cast from 'double *' to
'__attribute__((address_space(5))) double *' converts between
mismatching address spaces
      __x, (__attribute__((address_space(5))) double *)&__tmp);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cfs/klemming/root/rocm/opt/rocm-5.3.3/llvm/lib/clang/15.0.0/include/__clang_hip_math.h:1123:12:
 error: C-style cast from 'double *' to
'__attribute__((address_space(5))) double *' converts between
mismatching address spaces
      __x, (__attribute__((address_space(5))) double *)&__tmp);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10 errors generated when compiling for gfx90a.
make: *** [Makefile:48: main.o] Error 1
```

With mainline Clang 17 and SSCP:

```console
$ acpp -fsycl -O2 --acpp-targets=generic simple-sycl-app.cpp 
In file included from simple-sycl-app.cpp:1:
In file included from /home/aland/local/bin/../include/AdaptiveCpp/sycl/sycl.hpp:31:
In file included from /home/aland/local/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/sycl.hpp:58:
In file included from /home/aland/local/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/accessor.hpp:64:
/home/aland/local/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/detail/local_memory_allocator.hpp:205:35: error: C-style cast from '__attribute__((address_space(3))) void *' to 'char *' converts between mismatching address spaces
  205 |       return reinterpret_cast<T*>((char*)__hipsycl_sscp_get_dynamic_local_memory() + addr),
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/aland/local/bin/../include/AdaptiveCpp/sycl/../hipSYCL/sycl/libkernel/backend.hpp:166:5: note: expanded from macro '__hipsycl_backend_switch'
  166 |     sscp_code;                                                                 \
      |     ^~~~~~~~~
1 error generated.
```

**Review by illuhad — COMMENTED** [review]
Great idea! I would be fine with either a warning or error, but I would suggest rewording the warning slightly: To me it suggests that there is a deficiency on our side because we are incompatible with upstream LLVM.
I would phrase it in a less friendly way: The -fsycl flag is a component from a different (perhaps even competing?) SYCL implementation (DPC++) that was upstreamed into clang/LLVM with incomplete functionality. This flag is neither neither needed nor meaningful for AdaptiveCpp and its use is unsupported as it can actively break code.

Or something similar?

**Comment by al42and** [comment]
> Great idea! 

Not mine. There were some creative students in the course I was teaching :)

>  I would be fine with either a warning or error, but I would suggest rewording the warning slightly

How about making it a fatal error with the following, shorter, text?

The {} flag is a component from a different SYCL implementation. This flag is neither neither needed nor meaningful for AdaptiveCpp. Its use is unsupported.

**Comment by illuhad** [comment]
Awesome, we need more creative students :D

> How about making it a fatal error with the following, shorter, text?
> The {} flag is a component from a different SYCL implementation. This flag is neither neither needed nor meaningful for AdaptiveCpp. Its use is unsupported.

Fine for me!

**Review by illuhad — APPROVED** [review]
Thanks for implementing, and thanks to your students for the idea :)
